src/arch/src/aarch64/gic/dist_regs.rs

```diff
@@ -39,66 +39,73 @@ static VGIC_DIST_REGS: &[DistReg] = &[
-    length: u16,
+    offset: u32,
+    /// Distributor register size.
+    size: DistRegSize,
 }
 
 impl DistReg {
-    const fn new(base: u32, bpi: u8, length: u16) -> DistReg {
-        DistReg { base, bpi, length }
+    const fn simple(offset: u32, size: u16) -> DistReg {
+        DistReg {
+            offset,
+            size: DistRegSize::Simple { size },
+        }
     }
 
-    fn compute_reg_range(&self, fd: &DeviceFd) -> Result<(u32, u32)> {
-        // The ARM® TrustZone® implements a protection logic which contains a read-as-zero/write-ignore (RAZ/WI) policy
-        // where:
-        // * A blocked read operation will always return a zero value on the bus, preventing information leak
-        // * A write operation to a forbidden region or peripheral will be ignored
-        // The first part of a shared-irq type of register = (i.e GICD_<REG><0>) is RAZ/WI, so we compute the
-        // base by shifting it with "REG_SIZE * self.bpi" bytes.
-        let base = self.base + REG_SIZE as u32 * self.bpi as u32;
-        let mut end = base;
-        let num_irq = get_num_interrupts(fd)?;
-
-        if self.length > 0 {
-            // This is the single type register (i.e one that is not DIST_X<n>) and for which
-            // the bpi is 0.
-            // Look in the kernel for REGISTER_DESC_WITH_LENGTH.
-            end = base + self.length as u32;
+    const fn shared_irq(offset: u32, bits_per_irq: u8) -> DistReg {
+        DistReg {
+            offset,
+            size: DistRegSize::SharedIrq { bits_per_irq },
         }
-        if self.bpi > 0 {
-            // This is the type of register that takes into account the number of interrupts
-            // that the model has. It is also the type of register where
-            // a register relates to multiple interrupts.
-            end = base + (self.bpi as u32 * (num_irq - super::super::layout::IRQ_BASE) / 8);
-            if self.bpi as u32 * (num_irq - super::super::layout::IRQ_BASE) % 8 > 0 {
-                end += REG_SIZE as u32;
+    }
+}
+
+impl DistReg {
+    fn mem_iter(&self) -> impl Iterator<Item = u32> {
+        let mut start = self.offset;
+        let end = match self.size {
+            DistRegSize::Simple { size } => start + u32::from(size),
+            DistRegSize::SharedIrq { bits_per_irq } => {
+                // The ARM® TrustZone® implements a protection logic which contains a
+                // read-as-zero/write-ignore (RAZ/WI) policy.
+                // The first part of a shared-irq register, the one corresponding to the
+                // SGI and PPI IRQs (0-32) is RAZ/WI, so we skip it.
+                start += IRQ_BASE * u32::from(bits_per_irq) / 8;
+
+                let size_in_bits = u32::from(bits_per_irq) * (IRQ_MAX - IRQ_BASE);
+                let mut size_in_bytes = size_in_bits / 8;
+                if size_in_bits % 8 > 0 {
+                    size_in_bytes += u32::from(U32_SIZE);
+                }
```

Comment:
I am not sure how we should align these. If all regs are multiple of `U32_SIZE` and `size_in_bytes` is not a multiple of `U32_SIZE`, then the real end of the register should be `size_in_bytes` + (U32_SIZE - (size_in_bytes % U32_SIZE)). If we add u32::from(U32_SIZE) always we are going to enter in the RAZ/WI zone, right?

Reply:
I do not think I understand the concern. If there is a raz/wi zone, then that zone will be at the start of the register space. So, we could not possible reach it by adding bits to the the register space...What we do in this code is we try to read the amount of register space that is of interest to us (i.e the amount of space that covers the interrupts that our system uses). 
The GICD registers that share interrupts function in a slightly weird way. 
Let us say that we have 15 interrupts in our system, and there is a GICD register that uses 4 bits to represent some info about 1 interrupt. The way we detect the amount of register space is like this:
15 * 4 = 60 (our interrupts cover 60 bits of space)
a u32 has 32 bits =>  60 / 32 = 1 mod 28 => we will read 2 u32s from the register space to get our info on the 15 interrupts. Does it make sense?